pkg/store/proxy.go

```diff
@@ -181,6 +182,31 @@ func (s ctxRespSender) send(r *storepb.SeriesResponse) {
 // Series returns all series for a requested time range and label matcher. Requested series are taken from other
 // stores and proxied to RPC client. NOTE: Resulted data are not trimmed exactly to min and max time range.
 func (s *ProxyStore) Series(r *storepb.SeriesRequest, srv storepb.Store_SeriesServer) error {
+	// TODO(ppanyukov): hardcoded mem limit for now, where is best place to plug config for this?
+	const maxMem = uint64(2 * 1024 * 1024 * 1024) // 2GB
+
+	// checkMem returns an error if the current memory allocations exceeded max.
+	checkMem := func() error {
+		var nowMem uint64
+		{
+			// TODO(ppanyukov): ReadMemStats does stopTheWorld. Is this acceptable?
```

Comment:
I'm not sure there is a way to avoid it? Unless maybe we ask OS instead? Or we use completely our own allocator. Or the queries can tell us exactly how much memory they are consuming and freeing?

Reply:
qeuries?

Reply:
Yeah, if we have each query itself know know exactly how much memory they need, and tell some central accounting service when add acquire and release memory, we could do it.